File: backend/FitnessApp/src/Shared/Shared.Application/Common/Result.cs
using System.Net;
using Shared.Application.Abstractions;

namespace Shared.Application.Common;

public class Result<TResponse> : IResult<TResponse, Error>
{

    public TResponse? Response { get; }
    public Error? Error { get; }
    public bool IsSuccess { get; }
    public HttpStatusCode StatusCode { get; }

    private Result(TResponse? response)
    {
        Response = response;
        IsSuccess = true;
        StatusCode = HttpStatusCode.OK;
    }

    private Result(Error? error)
    {
        Error = error;
        IsSuccess = false;
        StatusCode = HttpStatusCode.BadRequest;
    }
    
    public static Result<TResponse> Success(TResponse response)
    {
        return new Result<TResponse>(response);
    }
    
    public static Result<TResponse> Failure(Error error)
    {
        return new Result<TResponse>(error);
    }
}

File: frontend/src/app/core/types/result/result.ts
import { Error } from "./error";

type SuccessResult<T> = {
  isSuccess: true;
  response: T;
  error?: undefined;
};

type FailureResult = {
  isSuccess: false;
  response?: undefined;
  error: Error;
};

export type Result<T> = SuccessResult<T> | FailureResult

export namespace Result {
  export function success<T>(response: T): Result<T> {
    return { isSuccess: true, response };
  }

  export function failure<T>(error: Error): Result<T> {
    return { isSuccess: false, error };
  }
}
